File: src/models/persistence.ts
import * as fs from 'fs'
import * as path from 'path'

const CHECKPOINT_DIR = path.join(process.cwd(), 'model-checkpoints')
const DATA_DIR = path.join(process.cwd(), 'training-data')

interface PersistedData {
  qTable: Record<string, Record<string, number>>
  experienceBuffer: {
    state: Record<string, number | boolean>
    action: { paramName: string; delta: number }
    reward: number
    nextState: Record<string, number | boolean>
    timestamp: number
  }[]
  trainingSteps: number
  version: number
  exportedAt: string
}

function ensureDir(dir: string): void {
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true })
  }
}

export function saveQTable(
  qTable: Record<string, Record<string, number>>,
  trainingSteps: number,
  label: string = 'latest'
): string {
  ensureDir(CHECKPOINT_DIR)

  const data: PersistedData = {
    qTable,
    experienceBuffer: [],
    trainingSteps,
    version: 1,
    exportedAt: new Date().toISOString(),
  }

  const filename = `qtable-${label}.json`
  const filepath = path.join(CHECKPOINT_DIR, filename)
  fs.writeFileSync(filepath, JSON.stringify(data, null, 2))

  const datedFilename = `qtable-${label}-${Date.now()}.json`
  fs.writeFileSync(
    path.join(CHECKPOINT_DIR, datedFilename),
    JSON.stringify(data, null, 2)
  )

  return filepath
}

export function loadLatestQTable(): PersistedData | null {
  ensureDir(CHECKPOINT_DIR)

  const files = fs.readdirSync(CHECKPOINT_DIR)
    .filter(f => f.startsWith('qtable-') && f.endsWith('.json'))
    .map(f => ({ name: f, mtime: fs.statSync(path.join(CHECKPOINT_DIR, f)).mtimeMs }))
    .sort((a, b) => b.mtime - a.mtime)

  if (files.length === 0) return null

  const latest = files[0].name
  try {
    const raw = fs.readFileSync(path.join(CHECKPOINT_DIR, latest), 'utf-8')
    const parsed = JSON.parse(raw)
    if (parsed.qTable) return parsed as PersistedData
    return { qTable: parsed, experienceBuffer: [], trainingSteps: 0, version: 1, exportedAt: '' }
  } catch {
    return null
  }
}

export function listCheckpoints(): { label: string; timestamp: string; size: number }[] {
  ensureDir(CHECKPOINT_DIR)

  return fs.readdirSync(CHECKPOINT_DIR)
    .filter(f => f.startsWith('qtable-') && f.endsWith('.json'))
    .map(f => {
      const fp = path.join(CHECKPOINT_DIR, f)
      const stat = fs.statSync(fp)
      return {
        label: f.replace(/^qtable-/, '').replace(/\.json$/, ''),
        timestamp: stat.mtime.toISOString(),
        size: stat.size,
      }
    })
    .sort((a, b) => new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime())
}

export function saveExperienceBuffer(
  experiences: any[],
  append: boolean = true
): string {
  ensureDir(DATA_DIR)

  const filename = `experiences-${Date.now()}.json`
  const filepath = path.join(DATA_DIR, filename)

  const data = experiences.map(exp => ({
    state: {
      repoStars: exp.state.repoStars,
      repoForks: exp.state.repoForks,
      fileCount: exp.state.fileCount,
      languageCount: exp.state.languageCount,
      readmeLength: exp.state.readmeLength,
      contributorCount: exp.state.contributorCount,
      hasTests: exp.state.hasTests,
      hasCI: exp.state.hasCI,
    },
    action: exp.action,
    reward: exp.reward,
    nextState: {
      repoStars: exp.nextState.repoStars,
      repoForks: exp.nextState.repoForks,
      fileCount: exp.nextState.fileCount,
      languageCount: exp.nextState.languageCount,
      readmeLength: exp.nextState.readmeLength,
      contributorCount: exp.nextState.contributorCount,
      hasTests: exp.nextState.hasTests,
      hasCI: exp.nextState.hasCI,
    },
    timestamp: exp.timestamp,
  }))

  if (append && fs.existsSync(filepath)) {
    const existing = JSON.parse(fs.readFileSync(filepath, 'utf-8'))
    data.push(...existing)
  }

  fs.writeFileSync(filepath, JSON.stringify(data, null, 2))
  return filepath
}

export function loadAllExperiences(): any[] {
  ensureDir(DATA_DIR)

  const files = fs.readdirSync(DATA_DIR)
    .filter(f => f.startsWith('experiences-') && f.endsWith('.json'))

  const all: any[] = []
  for (const file of files) {
    try {
      const raw = fs.readFileSync(path.join(DATA_DIR, file), 'utf-8')
      const data = JSON.parse(raw)
      if (Array.isArray(data)) all.push(...data)
    } catch {
    }
  }

  return all
}

export function getTrainingDataSize(): { files: number; experiences: number } {
  ensureDir(DATA_DIR)

  const files = fs.readdirSync(DATA_DIR)
    .filter(f => f.startsWith('experiences-') && f.endsWith('.json'))

  let totalExperiences = 0
  for (const file of files) {
    try {
      const raw = fs.readFileSync(path.join(DATA_DIR, file), 'utf-8')
      const data = JSON.parse(raw)
      if (Array.isArray(data)) totalExperiences += data.length
    } catch {
    }
  }

  return { files: files.length, experiences: totalExperiences }
}

File: backend/models/persistence.py
from __future__ import annotations
import json
import os
import time
from pathlib import Path
from ..config import CHECKPOINT_DIR, TRAINING_DIR


def _ensure_dir(d: Path) -> None:
    d.mkdir(parents=True, exist_ok=True)


def save_q_table(
    q_table: dict[str, dict[str, float]],
    training_steps: int,
    label: str = 'latest',
) -> str:
    _ensure_dir(CHECKPOINT_DIR)
    data = {
        'qTable': q_table,
        'experienceBuffer': [],
        'trainingSteps': training_steps,
        'version': 1,
        'exportedAt': _now_iso(),
    }
    filepath = CHECKPOINT_DIR / f'qtable-{label}.json'
    filepath.write_text(json.dumps(data, indent=2), encoding='utf-8')
    dated = CHECKPOINT_DIR / f'qtable-{label}-{int(time.time() * 1000)}.json'
    dated.write_text(json.dumps(data, indent=2), encoding='utf-8')
    return str(filepath)


def load_latest_q_table() -> dict | None:
    _ensure_dir(CHECKPOINT_DIR)
    files = sorted(
        [f for f in CHECKPOINT_DIR.iterdir() if f.name.startswith('qtable-') and f.suffix == '.json'],
        key=lambda f: f.stat().st_mtime, reverse=True,
    )
    if not files:
        return None
    try:
        raw = files[0].read_text(encoding='utf-8')
        parsed = json.loads(raw)
        if 'qTable' in parsed:
            return parsed
        return {'qTable': parsed, 'experienceBuffer': [], 'trainingSteps': 0, 'version': 1, 'exportedAt': ''}
    except (json.JSONDecodeError, OSError):
        return None


def list_checkpoints() -> list[dict]:
    _ensure_dir(CHECKPOINT_DIR)
    entries: list[dict] = []
    for f in CHECKPOINT_DIR.iterdir():
        if f.name.startswith('qtable-') and f.suffix == '.json':
            stat = f.stat()
            entries.append({
                'label': f.name.replace('qtable-', '').replace('.json', ''),
                'timestamp': _iso_from_ts(stat.st_mtime),
                'size': stat.st_size,
            })
    entries.sort(key=lambda x: x['timestamp'], reverse=True)
    return entries


def save_experience_buffer(experiences: list[dict], append: bool = True) -> str:
    _ensure_dir(TRAINING_DIR)
    filename = f'experiences-{int(time.time() * 1000)}.json'
    filepath = TRAINING_DIR / filename
    data = [
        {
            'state': {k: exp['state'][k] for k in ('repoStars', 'repoForks', 'fileCount', 'languageCount',
                       'readmeLength', 'contributorCount', 'hasTests', 'hasCI')},
            'action': exp['action'],
            'reward': exp['reward'],
            'nextState': {k: exp['nextState'][k] for k in ('repoStars', 'repoForks', 'fileCount', 'languageCount',
                           'readmeLength', 'contributorCount', 'hasTests', 'hasCI')},
            'timestamp': exp.get('timestamp', time.time() * 1000),
        }
        for exp in experiences
    ]
    if append and filepath.exists():
        existing = json.loads(filepath.read_text(encoding='utf-8'))
        data.extend(existing)
    filepath.write_text(json.dumps(data, indent=2), encoding='utf-8')
    return str(filepath)


def load_all_experiences() -> list[dict]:
    _ensure_dir(TRAINING_DIR)
    all_data: list[dict] = []
    for f in TRAINING_DIR.iterdir():
        if f.name.startswith('experiences-') and f.suffix == '.json':
            try:
                data = json.loads(f.read_text(encoding='utf-8'))
                if isinstance(data, list):
                    all_data.extend(data)
            except (json.JSONDecodeError, OSError):
                pass
    return all_data


def get_training_data_size() -> dict:
    _ensure_dir(TRAINING_DIR)
    files = [f for f in TRAINING_DIR.iterdir() if f.name.startswith('experiences-') and f.suffix == '.json']
    total = 0
    for f in files:
        try:
            data = json.loads(f.read_text(encoding='utf-8'))
            if isinstance(data, list):
                total += len(data)
        except (json.JSONDecodeError, OSError):
            pass
    return {'files': len(files), 'experiences': total}


def _now_iso() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()


def _iso_from_ts(ts: float) -> str:
    from datetime import datetime, timezone
    return datetime.fromtimestamp(ts, tz=timezone.utc).isoformat()
